### ISSUE TITLE
PyTorch low-precision YAT conv and embedding can return negative or infinite scores

### ISSUE THREAD

**Issue by mlnomadpy** [azettaai/nmn#54]
## Summary

PyTorch YAT convolution and embedding paths compute squared distance as norm sums minus twice the dot product without clamping cancellation errors to zero. Exact input/kernel matches in float16 or bfloat16 can therefore produce negative distances, negative scores, or infinity.

## Affected code

- `src/nmn/torch/layers/_yat_conv_core.py`
- `src/nmn/torch/embed.py`
- Forward and transpose convolution variants that share the core

## Reproduction

On commit `4f0298d` with PyTorch 2.11.0 CPU:

```python
import torch
from nmn.torch import YatConv1D, YatEmbed

for dtype, seed in [(torch.float16, 0), (torch.bfloat16, 9)]:
    torch.manual_seed(seed)
    conv = YatConv1D(
        1, 1, 3, bias=False, use_alpha=False,
        epsilon=1e-5, dtype=dtype,
    )
    conv_out = conv(conv.weight.detach().reshape(1, 1, 3))

    torch.manual_seed(seed)
    embed = YatEmbed(
        1, 3, use_alpha=False, epsilon=1e-5, dtype=dtype,
    )
    embed_out = embed.attend(embed.embedding.detach().reshape(1, 3))
    print(dtype, conv_out.item(), embed_out.item())
```

Observed:

```
torch.float16  -228.125   inf
torch.bfloat16 -2.984375 -97.5
```

For the float16 convolution, the computed distance is `-0.000244140625`; for the bfloat16 embedding it is `-0.015625`.

## Impact

Supported low-precision modes can silently violate the non-negative, finite YAT-score invariant and propagate invalid losses or gradients.

## Acceptance criteria

- [ ] Computed squared distances are clamped to at least zero before epsilon is added.
- [ ] The fix covers convolution, transpose convolution, and spherical/non-spherical embedding paths.
- [ ] CPU float16 and bfloat16 exact-match regression tests produce finite, non-negative outputs and finite gradients.
- [ ] Numerical parity against an fp32 reference is tested away from exact collisions.
